ecc/goldilocks/decaf.go

```diff
@@ -0,0 +1,137 @@
+type Decaf struct{ c twistCurve }
+
+// Elt is an element of the Decaf group. It must be always initialized using
+// one of the Decaf functions.
+type Elt struct{ p twistPoint }
+
+func (e Elt) String() string { return e.p.String() }
+
+// IsValid returns True if a is a valid element of the group.
+func (d Decaf) IsValid(a *Elt) bool { return d.c.IsOnCurve(&a.p) }
+
+// Identity returns the identity element of the group.
+func (d Decaf) Identity() *Elt { return &Elt{*d.c.Identity()} }
+
+// Generator returns the generator element of the group.
+func (d Decaf) Generator() *Elt { return &Elt{*d.c.pull(Curve{}.Generator())} }
+
+// Order returns a scalar with the order of the group.
+func (d Decaf) Order() Scalar { return order }
+
+// Add calculates c=a+b, where + is the group operation.
+func (d Decaf) Add(c, a, b *Elt) { q := a.p; q.Add(&b.p); c.p = q }
+
+// Double calculates c=a+a, where + is the group operation.
+func (d Decaf) Double(c, a *Elt) { c.p = a.p; c.p.Double() }
+
+// Neg calculates c=-a, where - is the inverse of the group operation.
+func (d Decaf) Neg(c, a *Elt) { c.p = a.p; c.p.cneg(1) }
+
+// Mul calculates c=n*a, where * is scalar multiplication on the group.
+func (d Decaf) Mul(c *Elt, n *Scalar, a *Elt) { d.c.ScalarMult(&c.p, n, &a.p) }
+
+// MulGen calculates c=n*g, where * is scalar multiplication on the group,
+// and g is the generator of the group.
+func (d Decaf) MulGen(c *Elt, n *Scalar) { d.c.ScalarBaseMult(&c.p, n) }
+
+// IsIdentity returns True if e is the identity of the group.
+func (e *Elt) IsIdentity() bool { return fp.IsZero(&e.p.x) }
+
+// IsEqual returns True if e=a, where = is an equivalence relation.
+func (e *Elt) IsEqual(a *Elt) bool {
+	l, r := &fp.Elt{}, &fp.Elt{}
+	fp.Mul(l, &e.p.x, &a.p.y)
+	fp.Mul(r, &a.p.x, &e.p.y)
+	fp.Sub(l, l, r)
+	return fp.IsZero(l)
+}
+
+// UnmarshalBinary if succeeds returns a Decaf element by decoding the first
+// DecafEncodingSize bytes of b.
+func (e *Elt) UnmarshalBinary(b []byte) error {
+	if len(b) < DecafEncodingSize {
+		return errInvalidDecoding
+	}
+
+	s := &fp.Elt{}
+	copy(s[:], b[:DecafEncodingSize])
+	isNeg := fp.Parity(s)
+	modulus := fp.P()
+	if isNeg == 1 || !isLessThan(b[:DecafEncodingSize], modulus[:]) {
+		return errInvalidDecoding
+	}
+
+	one, s2, den, num := &fp.Elt{}, &fp.Elt{}, &fp.Elt{}, &fp.Elt{}
+	isr, altx := &fp.Elt{}, &fp.Elt{}
+	t0, t1 := &fp.Elt{}, &fp.Elt{}
+	fp.SetOne(one)
+	fp.Sqr(s2, s)                    // s2  = s^2
+	fp.Sub(den, one, s2)             // den = 1 + a*s^2
+	fp.Mul(t1, s2, &paramDTwist)     // t1 = d*s^2
+	fp.Add(t1, t1, t1)               // t1 = 2*d*s^2
+	fp.Add(t1, t1, t1)               // t1 = 4*d*s^2
+	fp.Sqr(t0, den)                  // num = (1 + a*s^2)^2
+	fp.Sub(num, t0, t1)              // num = (1 + a*s^2)^2 - 4*d*s^2
+	fp.Mul(t0, t0, num)              // t0 = num*den^2
+	isQR := fp.InvSqrt(isr, one, t0) // v = 1/sqrt(num*den^2)
+	if !isQR {
+		return errInvalidDecoding
+	}
```

Comment:
This is not constant time.  You can perform the other operations and do a conditional move at the end.

Reply:
updated take a look